dbms/src/Storages/MergeTree/MergeTreeUniqueIndex.cpp

```diff
@@ -0,0 +1,411 @@
+        const auto & type = index.data_types[i];
+        auto new_column = type->createColumn();
+        type->deserializeBinaryBulk(*new_column, istr, rows_to_read, 0);
+
+        block.insert(ColumnWithTypeAndName(new_column->getPtr(), type, index.columns[i]));
+    }
+
+    set->insertFromBlock(block);
+}
+
+String MergeTreeUniqueGranule::toString() const
+{
+    String res = "";
+
+    const auto & columns = set->getSetElements();
+    for (size_t i = 0; i < index.columns.size(); ++i)
+    {
+        const auto & column = columns[i];
+        res += " [";
+        for (size_t j = 0; j < column->size(); ++j)
+        {
+            if (j != 0)
+                res += ", ";
+            Field field;
+            column->get(j, field);
+            res += applyVisitor(FieldVisitorToString(), field);
+        }
+        res += "]\n";
+    }
+
+    return res;
+}
+
+void MergeTreeUniqueGranule::update(const Block & new_block, size_t * pos, size_t limit)
+{
+    size_t rows_read = std::min(limit, new_block.rows() - *pos);
+
+    if (index.max_rows && size() > index.max_rows)
+    {
+        *pos += rows_read;
+        return;
+    }
+
+    Block key_block;
+    for (size_t i = 0; i < index.columns.size(); ++i)
+    {
+        const auto & name = index.columns[i];
+        const auto & type = index.data_types[i];
+        key_block.insert(
+                ColumnWithTypeAndName(
+                    new_block.getByName(name).column->cut(*pos, rows_read),
+                    type,
+                    name));
+    }
+
+    set->insertFromBlock(key_block);
+
+    *pos += rows_read;
+}
+
+Block MergeTreeUniqueGranule::getElementsBlock() const
+{
+    if (index.max_rows && size() > index.max_rows)
+        return index.header;
+    return index.header.cloneWithColumns(set->getSetElements());
+}
+
+
+UniqueCondition::UniqueCondition(
+        const SelectQueryInfo & query,
+        const Context & context,
+        const MergeTreeUniqueIndex &index)
+        : IndexCondition(), index(index)
+{
+    for (size_t i = 0, size = index.columns.size(); i < size; ++i)
+    {
+        std::string name = index.columns[i];
+        if (!key_columns.count(name))
+            key_columns.insert(name);
+    }
+
+    const ASTSelectQuery & select = typeid_cast<const ASTSelectQuery &>(*query.query);
+
+    /// Replace logical functions with bit functions.
+    /// Working with UInt8: last bit = can be true, previous = can be false.
+    ASTPtr new_expression;
+    if (select.where_expression && select.prewhere_expression)
+        new_expression = makeASTFunction(
+                "and",
+                select.where_expression->clone(),
+                select.prewhere_expression->clone());
+    else if (select.where_expression)
+        new_expression = select.where_expression->clone();
+    else if (select.prewhere_expression)
+        new_expression = select.prewhere_expression->clone();
+    else
+        /// 11_2 -- can be true and false at the same time
```

Comment:
11_2 - I guess you mean "11 in binary".
If you write 0b11 - it will be more obvious.